### PR TITLE
feat(floating-menu): add focus wrap element to components

### LINKED ISSUE #3603: [Floating Menu] Focus wrap should work without having to put in extra elements
@asudoh @dakahn @snidersd 

## What package(s) are you using?
- [x] `carbon-components`
- [ ] `carbon-components-react`

## Detailed description

> Describe in detail the issue you're having.

Demos are deceiving because there is a visually hidden input that is not part of the "demo" code but this input is required in order to have the a11y focus wrap feature work.  I understand that any "focusable" element following a floating menu would work and it does not have to be this specific input but we should not assume that of our users' use-cases.  Focus wrap should work without having to add in extra elements, especially if it's something like this input which is visually hidden but is visible to screen reader users, making it really confusing as to why they are reaching a text input.

> Is this issue related to a specific component?

Any component that uses `Floating Menu`

> What did you expect to happen? What happened instead? What would you like to
> see changed?

Focus wrap should work with the code provided on the website's demo code.  The focus wrap does not work if you copy and paste the code in to another system and even the Carbon code pen demo does not work as expected.

> What browser are you working in?

Chrome, Safari, Firefox

> What version of the Carbon Design System are you using?

v10.4.1

> What offering/product do you work on? Any pressing ship or release dates we
> should be aware of?

Watson Health Design Pattern & Asset Library; Release date coming up is end of Q3

## Steps to reproduce the issue

1. Open the overflow menu and navigate to the last item in the list
2. press `Tab`
3. focus is moved outside of the menu and menu remains open

> Please create a reduced test case in CodeSandbox

Doesn't work as expected
https://codepen.io/team/carbon/pen/PgNGop

Works once I add in extra incorrect element
https://codepen.io/elizabethsjudd/pen/mNmZdR

### PR DESCRIPTION
**Pull request by elizabethsjudd**
Closes #3603 

Move the focusable element used for the focus wrap feature to be within the component so developers don't have to add a hidden element to get this feature. This impacts modal, tooltip, and overflow menu.  Waiting on #3476 to be merged so I can test with the new functionality of tooltip.

#### Changelog

**Changed**

- floating menu now supports a wrapper node around the floating menu that contains a focusable span used to detect menu blur
- modal demo updated
- overflow menu and tooltip demos updated and vanilla JS updated to support the new wrapper element

#### Testing / Reviewing

Test overflow menu, modal, and tooltip keyboard interactions

### CODE REVIEW

File: packages/components/src/components/modal/modal.hbs
Comment (asudoh) [inline packages/components/src/components/modal/modal.hbs]: Why did you put additional `<div>` rather than simply adding focus trap element to the last? Note that no user-visible focusable element should be put outside `<div class="bx--modal-container">`.

File: packages/components/src/components/overflow-menu/overflow-menu.hbs
Comment (asudoh) [inline packages/components/src/components/overflow-menu/overflow-menu.hbs]: Similar question to modal - Would you put the focus trap element next to `</ul>` instead?

File: packages/components/src/components/overflow-menu/overflow-menu.hbs
Comment (asudoh) [inline packages/components/src/components/overflow-menu/overflow-menu.hbs]: Same question as another instance of overflow menu.

File: packages/components/src/components/tooltip/tooltip.hbs
Comment (asudoh) [inline packages/components/src/components/tooltip/tooltip.hbs]: Same question to overflow menu - Would you put the trap element next to `<div class="bx--tooltip__footer">`?

File: packages/components/src/components/tooltip/tooltip.hbs
Comment (asudoh) [inline packages/components/src/components/tooltip/tooltip.hbs]: Some question as another instance of tooltip body.

File: packages/components/src/components/tooltip/tooltip.hbs
Comment (asudoh) [inline packages/components/src/components/tooltip/tooltip.hbs]: Some question as another instance of tooltip body.

**Review by asudoh — COMMENTED** [review]
Thank you for coming up with a PR @elizabethsjudd!

**Comment by elizabethsjudd** [comment]
@asudoh here's the thought process behind my work
1. the focusable span needs to move with the floating menu so it's less DOM manipulation to grab a container that contains both the menu and the focusable tab vs having to grab both elements separately to move them
1. most frameworks require a root element so by not having a containing div for some components that created sibling elements at the root and we should keep consistent HTML between vanilla and the other JS frameworks (React, Angular, Vue, etc)
1. I thought the span has to be completely outside of the component (affected modal and tooltip)

**Comment by asudoh** [comment]
Thank you for explaining @elizabethsjudd. 1. makes sense to me and I'll withdraw my comments to overflow menu due to `<ul>`/`<li>` semantics.

Wrt. 2., you are right that Angular, etc. requires root element. Modal already has two-level containers, though - One (`<div class="bx--modal">`) works as the root and the backdrop, another (`<div class="bx--modal-container">`) works as the container of the content. Given we don't have content in backdrop area, putting hidden content in the backdrop area makes more sense. An example can be seen [here](https://github.com/carbon-design-system/carbon-custom-elements/pull/61/files#diff-7b81c66d6f3e9c4b82c40ee5ecd30e50R146).

Wrt 3., I guess you meant where the content is put in with your "in component" wording. If so, my response to 2. would address it.

File: packages/components/src/components/tooltip/tooltip.hbs
Comment (elizabethsjudd) [reply packages/components/src/components/tooltip/tooltip.hbs]: @asudoh the trap element needs to be outside of the content area of the tooltip not a sibling within it.  Currently with the way that the trap element works, it can not be a direct sibling of the content area which in the case of tooltips content is any direct child of `.bx--tooltip`

Current structure of tooltip
```
<bx--tooltip>
  <bx--tooltip__caret/>
  <p>Tooltip content</p>
  <bx--tooltip__footer/>
  <!-- trap element is a sibling of content if placed here and doesn't work -->
  <trap-element>
</bx--tooltip>
```

There are a couple different options:

**1. Scope the content area**

We could add a div within the tooltip around any of the content including the footer and update the event listener to watch the new div instead of the tooltip element

```
<bx--tooltip>
  <bx--tooltip__caret/>
  <div>
    <p>Tooltip content</p>
    <bx--tooltip__footer/>
  </div>
  <trap-element>
</bx--tooltip>
```

**2. Add a unique selector to the trap element**

We could give a unique selector to the trap element and watch for it to get focus and when it gets focus, immediate move it.  If that's the case we can keep the original HTML structure.

```
<bx--tooltip>
  <bx--tooltip__caret/>
  <p>Tooltip content</p>
  <bx--tooltip__footer/>
  <span tabindex="0" class="bx--focus-trap"></span>
</bx--tooltip>
```

Comment (asudoh) [reply packages/components/src/components/tooltip/tooltip.hbs]: My apologies, I seems to have missed this comment. I think 1. makes sense. BTW instead of wrapper node ref, let's move the root node up to the wrapper as above structure tells. If we need to refer to where the root node used to be, we can introduce `options.selectorContent`. Thanks!

**Comment by elizabethsjudd** [comment]
@asudoh I made changes to each of the components so the span is contained within the component and added the `selectorContent` feature to floating menus.

**Comment by elizabethsjudd** [comment]
@asudoh While I'm working on some of the more JS extensive issues for a11y, I feel like it might be easier for me to work in the evening/night so that we are on an overlapping schedule so that I can I can get feedback quicker. Right now the issue that I'm having is that I'm spending an hour or so a day working on tasks and then waiting for you to be able to respond. I recognize that with a 13 hour time difference this is expected but as I have more issues open I'm building more merge conflicts (which are never fun). I've already ran the idea of working one day a week with this different schedule with my team and they agree with this shift to help with communication and hopefully speed up the contribution process. Can we coordinate a day/times for me to do this schedule shift?

**Comment by asudoh** [comment]
The updated code looks basically good to me. Could you update tests? BTW thanks for ideas on collaboration, I pinged you offline for my ideas.

**Comment by elizabethsjudd** [comment]
@asudoh tests have been updated and ready for re-review

**Comment by asudoh** [comment]
Cool thanks @elizabethsjudd! Would you want to cover your enhancements in the test as well?

**Comment by elizabethsjudd** [comment]
@asudoh the tests for the focus wrap feature seem to already be in there because the HTML has the old hidden input.  Are you wanting me to test for the presence of the element or what?

**Review by asudoh — APPROVED** [review]
Oh I didn't realize that, sorry @elizabethsjudd!

**Comment by elizabethsjudd** [comment]
@carbon-design-system/developers Can I get another review on this PR so we can merge it?

**Comment by elizabethsjudd** [comment]
@carbon-design-system/developers This is still waiting on review.

**Comment by elizabethsjudd** [comment]
@tw15egan or @emyarod can one of you review this so it can get merged?  TIA

**Comment by jnm2377** [comment]
Will take a look at this! 

@elizabethsjudd can you update the PR title? It should follow our commit message conventions for our changelog. You can read more about this in our [developer handbook](https://github.com/carbon-design-system/carbon/blob/master/docs/developer-handbook.md#type).

**Review by tw15egan — APPROVED** [review]
LGTM -- sorry for delays!

**Review by dakahn — APPROVED** [review]
Thanks so much for being patient our new(ish) system for assigning PRs to reviewers still has a few kinks to work out.

**Comment by elizabethsjudd** [comment]
Thanks for taking a look @tw15egan and @dakahn!

**Pull request merged**
